go/ocr2/decryptionplugin/decryption.go

```diff
@@ -33,7 +33,7 @@ type decryptionPlugin struct {
 }
 
 // NewReportingPlugin complies with ReportingPluginFactory.
-func (f DecryptionReportingPluginFactory) NewReportingPlugin(rpConfig types.ReportingPluginConfig) (types.ReportingPlugin, types.ReportingPluginInfo, error) {
+func (f DecryptionReportingPluginFactory) NewReportingPlugin(ctx context.Context, rpConfig types.ReportingPluginConfig) (types.ReportingPlugin, types.ReportingPluginInfo, error) {
```

Comment:
Shouldn't this be `looppcontext types.LOOPPContext`? Otherwise I worry about a future maintainer of this code doing long-running work with one of these contexts.

The same concern would apply across all other repos and functions that take a `types.LOOPPContext`.

Reply:
I only settled for pattern under the impression that it would not leak out of libocr. I would like to re-emphasize that I think  `LOOPPContext` is the [wrong name](https://github.com/smartcontractkit/offchain-reporting/pull/482#issuecomment-1956511168). I think we need to take two steps back and re-evaluate the purpose of this type.
1. There is nothing about a regular context that implies doing something "long-running" is allowed or not. 
2. An implementer assuming that the presence of a context argument means that they can make a web request and block for 30 seconds is making a mistake. 
3. A method that has timing expectations needs to be clear about those expectations, regardless of whether or not a context is included in the argument set.
> Context type, which carries deadlines, cancellation signals, and other request-scoped values across API boundaries and between processes.

Reply:
> would like to re-emphasize that I think LOOPPContext is the wrong name

Thanks for re-iterating that you don't like the name, I'm happy to consider alternatives.
For now, I would suggest that we leave the naming aside and focus on the conceptual aspects.

I agree with your three points.
My concern is about misuse resistance. The intention behind `types.LOOPPContext` is to make people read the docs. 
Expanding on that (apologies if I am being repetitive): Contexts are frequently used for "long-running" interactions e.g. with remote servers. I worry that someone will (incorrectly as you point out, and without having read the docs on the interface) take the context and stuff it into whatever HTTP/database/Ethereum RPC call they want to perform, because they just assume that the presence of the context means that they're in "long-running" land. However, the context that's passed here effectively never expires (only when the underlying OCR configuration changes or the `offchainreporting2plus.Oracle` is shutdown). This is particularly concerning for the ReportingPlugin functions.

Reply:
(Lmk if you'd prefer to take this discussion back to the [original PR](https://github.com/smartcontractkit/offchain-reporting/pull/506) or elsewhere. Not sure whether this is the best place for it, since it goes beyond tdh2.)

Reply:
If there is widespread misunderstanding about context.Context implying long running network connections, then we need to correct that.

Reply:
> I worry that someone will (incorrectly as you point out, and without having read the docs on the interface) take the context and stuff it into whatever HTTP/Ethereum RPC call they want to perform, because they just assume that the presence of the context means that they're in "long-running" land. However, the context that's passed here effectively never expires (only when the underlying OCR configuration changes or the offchainreporting2plus.Oracle is shutdown). This is particularly concerning for the ReportingPlugin functions.

Using a different type of context (or even removing it all together) does not stop them from doing this, it just forces them to use a `context.Background()` with strictly _worse_ behavior due to being disconnected from any other contexts up the stack.

Reply:
@kaleofduty i am not convinced that argument naming can  prevent misunderstanding, bugs, or bad code.

since we all understand and accept that ctx, under one name or another, is required to meet our shared goals, would it better to spend our energy investing best practices and test suites for developers? Can we spend some time together and determine the behavior we expect from implementors as well as  how and where those expectations should be expressed? Perhaps a testing framework or a example implementations or well written godocs or something else.

At the end of the day, @jmank88 is correct that there is no way to prevent a developer from misusing the API, whether or not context is exposed. However, we obviously need some guardrails and we need your help to identify the most sensitive parts of the API, or the largest ROI guardrails.

Reply:
> If there is widespread misunderstanding about context.Context implying long running network connections, then we need to correct that.

A fair point ofc, but also orthogonal to the misuse resistance discussion.

Reply:
> A fair point ofc, but also orthogonal to the misuse resistance discussion.

I disagree. Basic understanding of the standing library is essential. Some level of developer competency has to be assumed and IMHO this is setting the bar far far too low.

Reply:
A new type is a non-starter. It will break tooling designed to recognize `context.Context` and litter our codebase with conversions back-and-forth between the two.

Reply:
It is not clear to me how a developer encountering this type is meant to do anything differently. I expect them to be confused initially, and then after reading the docs possibly confused more when they realize it is just a context with an ambiguously short runtime expectations of some kind. The correct thing to do is still just to meet the deadline, and the particulars of that deadline still depend on the method.

Reply:
> litter our codebase with conversions back-and-forth between the two

Isn't that kinda the point behind Kostis' suggestion? I would maybe phrase it a little more positively as "make conversions explicit to encourage developers to be conscious of semantics" 😉 

> It will break tooling

Would you mind elaborating on that?

Reply:
> > It will break tooling

> Would you mind elaborating on that?

Static checks and runtime tooling that look for the `context.Context` type will be rendered useless via false positives/negatives when unable to recognize our custom type.

Reply:
> Isn't that kinda the point behind Kostis' suggestion? I would maybe phrase it a little more positively as "make conversions explicit to encourage developers to be conscious of semantics" 😉

I also don't agree that an explicit conversion is a helpful thing. It doesn't provide the pertinent information, which is the actual timeout/deadline.

Reply:
@krehermann 

> i am not convinced that argument naming can prevent misunderstanding, bugs, or bad code.

Fully agree it cannot catch 100% of cases, but maybe help prevent some issues ( more from an ergonomics/devx perspective)? 
I think @gtklocker 's newtype suggestion would actually better than the alias , but it sounds like @jmank88 has some objections.

> since we all understand and accept that ctx, under one name or another, is required to meet our shared goals, would it better to spend our energy investing best practices and test suites for developers? Can we spend some time together and determine the behavior we expect from implementors as well as how and where those expectations should be expressed? Perhaps a testing framework or a example implementations or well written godocs or something else.

I hope that the docs around this on libocr are fairly comprehensive, but I would very much appreciate it if you spot any aspects on the ReportingPlugin (or the other injected deps that are affected here) that should be documented better and point them out to us!

As far as I can see, a testing framework would sadly be a in a different league of effort vs what's being discussed here.

Reply:
Rephrasing the tradeoff as I understand it for my understanding:
What amount of standard tooling breakage and type conversion boiler plate are we willing to introduce to signal to LibOCR consumers that they need to be careful with their usage of context? 

So, it seems like the best case scenario is one where 
* Don't break standard tooling
* Don't introduce unnecessary type conversion boiler plate
* Do signal to consumers that they need to be careful with their usage of context

Can we accomplish that best case scenario with standard Godoc comments?

Reply:
@jmank88 you wrote above

> and then after reading the docs possibly confused more when they realize it is just a context with ambiguously short runtime expectations of some kind. 

Am I reading this correctly that you are unhappy with the docs? But then why ask to merge the original PR?

> The correct thing to do is still just to meet the deadline

But many of these will not have a deadline!?

Reply:
> Am I reading this correctly that you are unhappy with the docs? But then why ask to merge the original PR?

The whole approach has diverged from my proposal and feedback. I have tried to be maximally receptive to this alternative in the interest of time and with the understanding that it would not have cascading effects across other repos.

> But many of these will not have a deadline!? They only expire once offchainreporting2plus.Oracle.Close() is called or there is a setConfig on the contract.

Should they have a deadline? It seems in conflict to claim they should not have a deadline, but also that the implementations "need to be fast". Why can't/shouldn't "fast" be quantified as specific deadlines?

Reply:
> Should they have a deadline? It seems in conflict to claim they should not have a deadline, but also that the implementations "need to be fast". Why can't/shouldn't "fast" be quantified as specific deadlines?

Adding a good deadline for these is tricky for a few reasons, here are some of them:
- it would force reporting plugin implementers to think about more timing parameters. setting the params wrong risks breaking things.
- timing parameters for the other reportingplugin functions are set via setConfig, adding more parameters there is a far reaching change!

Reply:
I think we need to be precise about the problem being solved.

If the author of the interface and caller of the method are not equipped to handle this, then how is an 
implementer supposed to be any better equipped to solve the problem?

I am not proposing adding a bunch of new fine tuned configuration. I would propose that we start by doing the obvious naive thing, which is to not block forever. e.g. If the upstream caller knows a round has ended, then it should cancel all associated contexts, not pass an unbounded one that may block forever. Only after we have exhausted the standard patterns for context usage should we start trying to invent new complexities. It  has not been stated clearly where the idiomatic approaches fall short.